Implement nav by find box functionality

jackjyang opened the pull request:
This allows navigating the timeline by entering a string in
the find input box in the format of (timestamp)@(pid).(tid)x(scaleX).
A line will be drawn at that ts, the pid and tid will be brought into
view, and zoom will be applied.

The find string can also be generated by double clicking the timeline
while in mousemode=selection. A nav line will be drawn and the nav string
will be generated in the find box to be copied by the user. Single clicking
anywhere on the timeline dismisses the nav line and removes the string.

This will be useful when one would want to refer to a specific point
on the timeline in - for example - an email thread.

jackjyang commented:
Posting this CL just as an fyi to @natduca @dj2 since we're planning out traceviewer annotations and ui states. Maybe some of this can be reused.

This is a work in progress -- a way to generate the string for easier copying will be implemented later.

jackjyang commented:
I've finally updated this - please take a look @natduca @dj2. I honestly think it's a little messy how I'm implementing this right now but I'm out of ideas on how to clean it up. Suggestions are always welcome!

dj2 commented on trace_viewer/core/find_controller.html:
```diff
@@ -63,18 +64,38 @@
-        }.bind(this));
+      var navByFinderPattern = /^(\d+(\.\d+)?)@(\d+).(\d+)x(\d+(\.\d+)?)$/g;
+      var match = navByFinderPattern.exec(this.filterText);
+      if (match != null) {
```
nil: !==

dj2 commented on trace_viewer/core/find_controller.html:
```diff
@@ -63,18 +64,38 @@
+      var navByFinderPattern = /^(\d+(\.\d+)?)@(\d+).(\d+)x(\d+(\.\d+)?)$/g;
+      var match = navByFinderPattern.exec(this.filterText);
+      if (match != null) {
+        var timestamp = match[1] * 1; // Convert string to a number.
```
var timestamp = parseInt(match[1]);

Doesn't need the comment in this case.

dj2 commented on trace_viewer/core/find_controller.html:
```diff
@@ -63,18 +64,38 @@
+            this.filterHits_ = filterHits;
+            this.timeline.setHighlightAndClearSelection(filterHits);
+          }.bind(this));
+        }
```
Can we break this into 3 methods to make it a bit clearer?

if (match === null) {
  if (this.filterText.length ===) {
    clearNavigationMarker();
  } else if (this.timeline_) {
    find() ? (not sure what the righ tname is;
  }
} else {
  setNavigationMarker();
}

Something like that?

dj2 commented on trace_viewer/core/location.html:
```diff
@@ -67,11 +87,33 @@
       }
     },
 
+    fromStableIdAndTs: function(stableId, ts) {
```
nit: Make this fromStableIdAndTimestamp  (My initial though was, what are the T's?)

dj2 commented on trace_viewer/core/location.html:
```diff
@@ -67,11 +87,33 @@
+      this.yComponents_ = [];
+      var containerToTrack = this.viewport_.containerToTrackObj;
+      var elem = containerToTrack.getTrackByStableId(stableId);
+      if (elem) {
```
if (!elem)
  return;

dj2 commented on trace_viewer/core/timeline_track_view.html:
```diff
@@ -479,11 +482,41 @@
+
+      // Remove nav line if created from double click. Otherwise it will be
+      // removed once find text is cleared.
+      if (this.modelTrack_.navLineTs >= 0 && this.navLineAtStableId_) {
```
nit: change navLineTs to navLineTimestamps

jackjyang commented:
I've elected to use Number() instead of parseInt() since it handles conversion of decimal numbers as well. But I addressed all other comments. Thanks!

dj2 commented:
This seems good to me as a first pass. @natduca do you have any issues?

natduca commented:
Pretty swamped so lets land when you're happy Dan :) Sorry I held this up. :/

jackjyang commented:
@natduca's earlier comment that was lost when I pushed:
> actually this bit is a bit hacky... lets maybe do the pieces of this patch that are clear [the location etc glue] and make a new button tool in the mouse mode selector for making an annotation.... as a separate patch

So I've updated this cl with only the parts responsible for handling typing in find box, parsing, bringing into view and drawing a line.

On second thought I think it's worth having more discussion on how this second part will work. That is, the part to generate the find string so that it's easily copyable and sharable. I think the whole thing will still be a little 'hacky' without a generic way to store and navigate to the current state of the timeline. Nat I noticed you were looking into TraceViewer UI states a while ago -- I think this is something that will significantly overlap with this feature. I just fear that this code might need to be redone once we have actual UI states because even if we create a new mouse mode for this find string generation part the underlying functionality might be duplicated here.

jackjyang commented:
Hi @natduca, I've updated this PR with a new fix-up commit -- I didn't want to lose your comments. Please take a look. After you lgtm if you want I can squash everything into a single commit before merging.

Also since I removed my commit about drawing the line, Github lost our comments on that. I'm copying that here, mainly for my reference when adding annotations:

>Eek that we're duplicating this code twice here. Gotta find a way to avoid that.
We should be finding an abstraction here that can capture your use case but not duplicate code.
Suppose we make an annotation class. This is a virtual base class that has draw methods just like tracks do. So its base class is constructed with a viewport argument.
And, suppose we have Viewport have a list of annotations. E.g. viewport.addAnnotation, viewport.removeAnnotation. When you add/remove annotations, you call invalidate() like the tracks do, in order to cause a redraw.
WHich is awesome because suppose that the drawing container iterates through the viewport's active annotations every time it draws.
Now you can make a GlobalVerticalXMarker annotation that contains your drawing code that you construct from a single timestamp.
Now you can have the find controller construct a marker, add it to the viewport. The only thing that remains is asking the timeline_view to focus a location. So then al you need on the timeline is a scrollLocationIntoView(location, opt_dontAnimate).
Make sense?

And what I replied:
> Yup, thanks, that makes sense. I'll put off this part of this feature (ie. the marker drawing at timestamp) until we have some sort of annotation drawing capability. In the meantime I'll update this pull request just to commit the first part of the change once I address all your other comments.

jackjyang commented on trace_viewer/core/find_control.html:
```diff
@@ -112,7 +112,6 @@
 
     filterFocus: function(e) {
       this.controller.reset();
-      this.filterTextChanged();
```
I removed this because when a user inputs an unaccepted ui state string, an overlay is invoked and focus is no longer in the find box. Once the user dismisses the overlay and focuses back on the find box, the overlay is incorrectly invoked again because of this function even though the text was exactly the same as before.

I'm a little unsure about this removal -- as far as I can tell this shouldn't be there, and I couldn't find anything that broke after I removed it..

natduca commented:
looking solid. i think we need to squash these two commits down and move to retiveld. since its only two commits now i think its okay to have them in one. reitveld does comment preservation way better, anyway.

jackjyang commented:
Thanks Nat! I've fixed all the issues you've pointed out and uploaded the latest version on Rietveld: https://codereview.appspot.com/210660043/
